src/typecheck/eq.rs

```diff
@@ -0,0 +1,356 @@
+        }
+        (RecRecord(m1, dyn_fields1, attrs1, _), RecRecord(m2, dyn_fields2, attrs2, _)) =>
+        // We only compare records which field structure is statically known (i.e. without
+        // dynamic fields).
+        {
+            dyn_fields1.is_empty()
+                && dyn_fields2.is_empty()
+                && map_eq(contract_eq_bounded, state, m1, env1, m2, env2)
+                && attrs1 == attrs2
+        }
+        (Array(ts1, attrs1), Array(ts2, attrs2)) => {
+            ts1.len() == ts2.len()
+                && ts1
+                    .iter()
+                    .zip(ts2.iter())
+                    .all(|(t1, t2)| contract_eq_bounded(state, t1, env1, t2, env2))
+                && attrs1 == attrs2
+        }
+        // We must compare the inner values as well as the corresponding contracts or type
+        // annotations. Documentation and merge priority shouldn't impact the result on the
+        // other hand.
+        (MetaValue(m1), MetaValue(m2)) => {
+            let value_eq = match (&m1.value, &m2.value) {
+                (None, None) => true,
+                (Some(v1), Some(v2)) => contract_eq_bounded(state, v1, env1, v2, env2),
+                _ => false,
+            };
+
+            // We use the same logic as in the typechecker: the type associated to an annotated
+            // value is either the type annotation, or the first contract annotation.
+            let ty1 = m1.types.as_ref().or(m1.contracts.first());
+            let ty2 = m2.types.as_ref().or(m2.contracts.first());
+
+            let ty_eq = match (ty1, ty2) {
+                (None, None) => true,
+                (Some(ctr1), Some(ctr2)) => type_eq_bounded(
+                    state,
+                    &TypeWrapper::from(ctr1.types.clone()),
+                    env1,
+                    &TypeWrapper::from(ctr2.types.clone()),
+                    env2,
+                ),
+                _ => false,
+            };
+
+            value_eq && ty_eq
+        }
+        // We dont treat imports and parse errors
+        _ => false,
+    }
+}
+
+/// Compute the equality between two hashmaps holding either types or terms.
+fn map_eq<V, F>(
+    mut f: F,
+    state: &mut State,
+    map1: &HashMap<Ident, V>,
+    env1: &TermEnvironment,
+    map2: &HashMap<Ident, V>,
+    env2: &TermEnvironment,
+) -> bool
+where
+    F: FnMut(&mut State, &V, &TermEnvironment, &V, &TermEnvironment) -> bool,
+{
+    map1.len() == map2.len()
+        && map1.iter().all(|(id, v1)| {
+            map2.get(id)
+                .map(|v2| f(state, v1, env1, v2, env2))
+                .unwrap_or(false)
+        })
+}
+
+/// Convert rows (represented as linked list for the needs of unification) to a hashmap to perform
+/// the type equality comparison.
+///
+/// Require the rows to be closed (i.e. the last element must be RowEmpty), otherwise `None` is
+/// returned. `None` is returned as well if a type encountered is not a row type, a if it is a enum
```

Comment:
Not really, because I think this function is called only for rows of a record. However, you made me realize I didn't treat enums properly, and the current version probably requires all the tags to be the same in the exact same order...